ext/drx/drx.c

```diff
@@ -2429,3 +2441,363 @@ drx_expand_scatter_gather(void *drcontext, instrlist_t *bb, OUT bool *expanded)
+static inline bool
+allow_for_unknown_instr_inc(int *allow_for_unknown_instr_count)
+{
+    if (*allow_for_unknown_instr_count++ < DRX_RESTORE_EVENT_ALLOW_FOR_UNKNOWN_INSTR)
+        return true;
+    return false;
+}
+
+static void
+advance_state(int *detect_state, int new_detect_state, int *allow_for_unknown_instr_count)
+{
+    *detect_state = new_detect_state;
+    *allow_for_unknown_instr_count = 0;
+}
+
+static void
+drx_try_to_detect_avx512_gather_sequence(void *drcontext, dr_restore_state_info_t *info,
+                                         instr_t *inst, scatter_gather_info_t *sg_info)
+{
+    byte *prev_pc, *pc;
+    byte *restore_dest_mask_start = NULL;
+    byte *restore_scratch_mask_start = NULL;
+    int detect_state = 0;
+    int allow_for_unknown_instr_count = 0;
+    reg_id_t tmp_xmm = DR_REG_NULL;
+    reg_id_t gpr_bit_mask = DR_REG_NULL;
+    reg_id_t gpr_save_scratch_mask = DR_REG_NULL;
+    uint scalar_mask_update_no = 0;
+    pc = info->fragment_info.cache_start_pc;
+    /* As the state machine is looking for blocks of code that the fault may hit, the 128
+     * bytes is a conservative approximation of the block's size, see (a) and (b) above.
+     */
+    while (pc <= info->raw_mcontext->pc + 128) {
+        instr_reset(drcontext, inst);
+        prev_pc = pc;
+        pc = decode(drcontext, pc, inst);
+        if (pc == NULL) {
+            /* Upon a decoding error we simply give up. */
+            return;
+        }
+        /* If there is a gather or scatter instruction in the code cache, then it is wise
+         * to assume that this is not an emulated sequence that we need to examine
+         * further.
+         */
+        if (instr_is_gather(inst))
+            return;
+        if (instr_is_scatter(inst))
+            return;
+        switch (detect_state) {
+        case DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0:
+            if (instr_reads_memory(inst)) {
+                opnd_t dst0 = instr_get_dst(inst, 0);
+                if (opnd_is_reg(dst0) && reg_is_gpr(opnd_get_reg(dst0))) {
+                    restore_dest_mask_start = pc;
+                    advance_state(&detect_state,
+                                  DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_1,
+                                  &allow_for_unknown_instr_count);
+                    break;
+                }
+            }
+            break;
+        case DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_1:
+            if (instr_get_opcode(inst) == OP_vextracti32x4) {
+                reg_id_t tmp_reg = opnd_get_reg(instr_get_dst(inst, 0));
+                if (!reg_is_strictly_xmm(tmp_reg))
+                    break;
+                tmp_xmm = tmp_reg;
+                advance_state(&detect_state,
+                              DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_2,
+                              &allow_for_unknown_instr_count);
+                break;
+            }
+            /* Intentionally not else if */
```

Comment:
OK, but it's a little weird to have a *known* instr count toward an *unknown* instr limit: maybe rename that limit.

Reply:
s/allow_for_unknown_instr_count/unknown_instr_count/
Better?

Reply:
No that makes it sound even more like there could not be any known instrs in there.  I guess the original is ok.  But what about sthg like "sequence_instr_count" which seems clearer to me.

Reply:
But it's not sequence_instr_count. It's the counter for instruction _in between_ the sequence instructions, that we allow for and ignore. For example if somebody had added instrumentation => in expectation of memory instrumentation clients for example.

Reply:
Then a name that describes that.  If you change the code to *not* count the known instr that may make it easier.

Reply:
What about skip_unknown_instr_count?

Reply:
And s/DRX_RESTORE_EVENT_ALLOW_FOR_UNKNOWN_INSTR/DRX_RESTORE_EVENT_SKIP_FOR_UNKNOWN_INSTR_MAX/

Reply:
And s/allow_for_unknown_instr_inc/skip_unknown_instr_inc/

Reply:
OK, w/o the known instrs in the count it makes sense.